### PR TITLE
Add head_mask, decoder_head_mask, cross_head_mask to ProphetNet

### LINKED ISSUE #9814: Missing head_mask and decoder_head_mask arguments in encoder-decoder models
# 🚀 Feature request

Following the PRs #9569, #9634 and #9639, there are other encoder-decoder models, which either do not support `head_mask` and `decoder_head_mask` input arguments at all or can be only provided with a single `head_mask` argument used for head masking both in encoder and decoder. It would be, therefore, nice to make this feature uniform over all the decoder-models.

<hr>

**Models:**

| Model | Pytorch | TensorFlow | PR | Copy dependency |
| ------ | :------: | :---------: | :--: | :-----: |
| BERTGeneration | ☑️  | ✖️  | - | - |
| EncoderDecoderModel | ☑️  | ✖️  | - | - |
| FSMT | ☑️  | ✖️  | #9819 | - |
| LED | ☑️ | ☑️ | PT - #9856 ; TF | - |
| ProphetNet | ☑️  | ✖️  | - | - |
| Longformer | ☑️  | ✖️  | #9856  | LED |

## Your contribution

I'm happy to add this feature in the following days, both for PyTorch and TensorFlow models. (Likely in shorter PRs in order not to create large, overwhelming PRs)

<hr>

Reviewers: @patrickvonplaten, @jplu, @sgugger, @LysandreJik, @stas00 .

### PR DESCRIPTION
**Pull request by stancld**
This PR implements `head_mask`, `decoder_head_mask` and `cross_head_mask` for ProphetNet (and Longformer as there's a copy dependency) and it is the follow-up to the open issue #9814. 

**Motivation:** This PR is a part of an endeavour to enable the usage of `head_mask` and `decoder_head_mask` for all encoder-decoder transformers following the recent work on BART-like models (#9569).

<hr>

Fixes: https://github.com/huggingface/transformers/issues/9814

Reviewers: @patrickvonplaten

### CODE REVIEW

**Comment by LysandreJik** [comment]
Reopened as this might still be in the works.

**Comment by stancld** [comment]
This PR is required for #10605.

Also, it is necessary to rebase this branch to the current `master` [As a lot of changes have been done to the repo, there are some conflicts I'm gonna handle asap.].

**Comment by stancld** [comment]
Hi @LysandreJik - I fixed `cross_head_mask` for this `ProphetNetModel`. At this moment, there is an error regarding the `test_forward_signature` and there is likely to be a problem with a template. This issue should be then resolved in #10605 which takes care of `cross_head_mask` for all other encoder-decoder models which have already had `head_mask` and `decoder_head_mask` merged into the master.

**Comment by stancld** [comment]
Update: #10605 now passes all the tests. (@LysandreJik)

**Comment by patrickvonplaten** [comment]
Super sorry for being incredibly slow here @stancld ! I think we can actually merge this if it passes all the tests :-)

**Comment by stancld** [comment]
@patrickvonplaten No worries, it's completely okay! :) I rebase this branch and now all the tests have passed.

**Review by patrickvonplaten — APPROVED** [review]
Great!
